src/raw/mod.rs

```diff
@@ -0,0 +1,1634 @@
+//! ```
+
+mod elem;
+#[cfg(test)]
+mod props;
+
+use std::{
+    borrow::Borrow,
+    convert::{TryFrom, TryInto},
+    ops::Deref,
+};
+
+use chrono::{DateTime, Utc};
+
+#[cfg(feature = "decimal128")]
+use crate::decimal128::Decimal128;
+use crate::{oid::ObjectId, spec::ElementType, Bson, Document};
+pub use elem::{RawBinary, RawBson, RawRegex, RawTimestamp};
+
+/// An error that occurs when attempting to parse raw BSON bytes.
+#[derive(Debug, PartialEq)]
+pub enum RawError {
+    /// A BSON value did not fit the expected type.
+    UnexpectedType,
+
+    /// A BSON value did not fit the proper format.
+    MalformedValue(String),
+
+    /// Improper UTF-8 bytes were found when proper UTF-7 was expected. The error value contains
+    /// the malformed data as bytes.
+    Utf8EncodingError(Vec<u8>),
+}
+
+impl std::fmt::Display for RawError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::UnexpectedType => write!(f, "unexpected type"),
+            Self::MalformedValue(s) => write!(f, "malformed value: {:?}", s),
+            Self::Utf8EncodingError(_) => write!(f, "utf-8 encoding error"),
+        }
+    }
+}
+
+impl std::error::Error for RawError {}
+
+pub type RawResult<T> = Result<T, RawError>;
+
+/// A BSON document, stored as raw bytes on the heap. This can be created from a `Vec<u8>` or
+/// a [`bson::Document`].
+///
+/// Accessing elements within a `RawDocument` is similar to element access in [bson::Document], but
+/// because the contents are parsed during iteration, instead of at creation time, format errors can
+/// happen at any time during use.
+///
+/// Iterating over a RawDocument yields either an error or a key-value pair that borrows from the
+/// original document without making any additional allocations.
+///
+/// ```
+/// # use bson::raw::{RawDocument, RawError};
+/// let doc = RawDocument::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+/// let mut iter = doc.iter();
+/// let (key, value) = iter.next().unwrap()?;
+/// assert_eq!(key, "hi");
+/// assert_eq!(value.as_str(), Ok("y'all"));
+/// assert!(iter.next().is_none());
+/// # Ok::<(), RawError>(())
+/// ```
+///
+/// Individual elements can be accessed using [`RawDocument::get`](RawDocument::get) or any of the
+/// type-specific getters, such as [`RawDocument::get_object_id`](RawDocument::get_object_id) or
+/// [`RawDocument::get_str`](RawDocument::get_str). Note that accessing elements is an O(N)
+/// operation, as it requires iterating through the document from the beginning to find the
+/// requested key.
+///
+/// ```
+/// # use bson::raw::{RawDocument, RawError};
+/// let doc = RawDocument::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+/// assert_eq!(doc.get_str("hi")?, Some("y'all"));
+/// # Ok::<(), RawError>(())
+/// ```
+#[derive(Clone, Debug)]
+pub struct RawDocument {
+    data: Box<[u8]>,
+}
+
+impl RawDocument {
```

Comment:
I think we'll want some methods that support appending to this type, otherwise users will have to route through `Document`, negating many of the performance benefits of sticking with raw.

Reply:
I had started some work on this that I guess never made it in.  One thing to note is that keeping the methods performant necessitates not checking for duplicate keys, since finding a key is O(N) on the length of the document.  Similarly for finding the length of a RawArray (which is also a big ergonomic hit, since `array.append(x)` becomes `array.insert(the_right_index, x)`.  One option for arrays would be a RawArrayAppender type, that lets you append multiple entries while only checking the current length once, when the appender is created.  I don't love that either, though.